editorial pass on section 3 (path management) (#507)
quicwg/multipath: draft-ietf-quic-multipath.md

# Path Management {#path-management}

> After completing the handshake indicating
> multipath support, endpoints can start using multiple paths when both endpoints
> have issued available connection IDs for at least one unused, common Path ID,
> as the same Path ID is used in both directions.

Author: ```suggestion
After completing the handshake indicating
multipath support, endpoints can start using multiple paths.
An endpoint can open a new path when both endpoints
have issued available connection IDs for at least one unused, common Path ID,
as the same Path ID is used in both directions.
```
Is this better?

Reviewer: Yes.

> 

Reviewer: Not sure that this rewrite is better than the original. You could just strike out the last line, "This document does not discuss when a client decides to initiate a new path."

Author: The initial text is not correct. 

"After completing the handshake, endpoints have agreed to enable
multipath support." -> only if both ends indicated multipath support (but I don't think we need to repeat this here.

I think this happened because this used to be use sentence we split it up (wrongly) at some point.

Please see another proposal to split this into two sentence above.

Reviewer: New proposal works.

This documents specfies path initiation (see Section {{path-initiation}}),
issuing and retirement of per-path connection IDs (see Section
{{consume-retire-cid}}), path status management (see Section {{path-state}}) and

> path closure (see Section {{path-close}}).

Reviewer: +1

However, this document does not specify when a client decides to initiate or close a path,

> or how multiple active paths are used for packet sending.

Author: ```suggestion
or how multiple active paths are used for sending.
```

> 

Reviewer: "are used for sending packets." ?

## Path Initiation and Validation {#path-initiation}

> To open a new path, an endpoint MUST use a new connection ID associated

Author: ```suggestion
to open a new path, an endpoint MUST use a new connection ID associated
```

> with unused Path ID.

Reviewer 2: ```suggestion
with an unused Path ID.
```

Reviewer: Please also fix the upper case "T" in "To open"

When sending a PATH_RESPONSE frame, an endpoint MUST use a connection ID associated to

> the same Path ID as used in packet that contained the PATH_CHALLENGE frame.

Reviewer 2: ```suggestion
the same Path ID as used in the packet that contained the PATH_CHALLENGE frame.
```

applies.

The server may receive packets for a yet unused Path ID that do not
contain a PATH_CHALLENGE frame. Such packets are valid if they can be properly decrypted
based on a valid connection ID.

Each endpoint MUST also validate that a minimum QUIC packet MTU of 1200 bytes is supported
on the path. This can be done during initial path validation or separately later if
the amplification limit prevents it initially, as specified in {{Section 8.2.1 of QUIC-TRANSPORT}}.

An endpoint that receives packets on a new path and does not want to establish
this path is expected to close the path by sending a PATH_ABANDON

> on another path, as specified in Section {{path-close}}.

Reviewer 2: ```suggestion
on another path, as specified in {{path-close}}.
```

applies.

The server may receive packets for a yet unused Path ID that do not
contain a PATH_CHALLENGE frame. Such packets are valid if they can be properly decrypted

> based on a valid connection ID.

Reviewer 3: The new sentence is not correct: AEAD decryption is not based on connection ID, but based on a Path ID as part of nonce. The original text is correct on logic, but hard to read. I'd suggest the new text changes to:

```suggestion
and contain a valid Connection ID.
```

Reviewer 2: The "based on" does not bother me, but the "given" version is maybe less ambiguous -- both of these versions works. I would not take the "and contain" as it would alter the original message of this sentence.

send a PATH_ABANDON, in which case the procedures specified in
Section {{path-close}} apply.

> Any frames can be sent on any open path.

Reviewer: Previous version is better. PATH_CHALLENGE and PATH_RESPONSE frames cannot be "sent on any open path".

Author: The point is PATH_CHALLENGE and PATH_RESPONSE frames are not special. There are no restrictions for any frame on which path to sent. We don't say this anywhere. I think it good to say that once somewhere.

Reviewer: They are special. You cannot send a challenge or a response for path 1 on path2...

Reviewer: I think we should keep the old text, for the same reason that we have text stating that "PATH_ABANDON frames can be sent on any path, not only on the path that is intended to be closed." It is important to assert that loudly, so implementers do not start assuming that PATH_ACK can only be sent on the path on which the acked packets were sent.

Author: Okay you are right. Reverting it.

Author: ```suggestion
PATH_ACK frames (see {{mp-ack-frame}}) can be returned on any path.
```

If the PATH_ACK is preferred to be sent on the same path as the acknowledged

> packet (see Section {{compute-rtt}} for further guidance), it can be beneficial

Reviewer 2: ```suggestion
packet (see {{compute-rtt}} for further guidance), it can be beneficial
```

to bundle a PATH_ACK frame with the PATH_RESPONSE frame during
path validation.

If validation succeeds, the client can continue to use the path.
If validation fails, the client MUST NOT use the path and can
remove any status associated to the path initiation attempt.
As the used Path ID is anyway consumed,
the endpoint MUST explicitly close the path, as specified in

> Section {{path-close}}.

Reviewer 2: ```suggestion
{{path-close}}.
```

Reviewer: +1

The client is likely to fall back to the strategy specified in {{Section 8.1.3 of QUIC-TRANSPORT}},
i.e., pick the last received token. To avoid issues when clients make the "wrong" choice,
a server SHOULD issue tokens that are capable of validating
any of the previously validated addresses. Further guidance on token usage can be
found in {{Section 8.1.3 of QUIC-TRANSPORT}}.

## Handling Connection IDs {#consume-retire-cid}

In order to identify a connection ID correctly when the multipath extension is used,
endpoints have to use the PATH_NEW_CONNECTION_ID and PATH_RETIRE_CONNECTION_ID frames
instead of the NEW_CONNECTION_ID and RETIRE_CONNECTION_ID frame to indicate
the respective Path ID together with associated sequence number
(see {{Section 5.1.1 of QUIC-TRANSPORT}}), at least for all paths with a Path ID other than 0.
Each Path ID has its own connection ID sequence number space whose initial value is 0.

Endpoints can also use PATH_NEW_CONNECTION_ID and
PATH_RETIRE_CONNECTION_ID for the initial path with Path ID 0,
however, the use of NEW_CONNECTION_ID and RETIRE_CONNECTION_ID
is still valid and endpoints need to process these frames accordingly

> as corresponding to Path ID 0.

Reviewer 2: Note regarding other editorial passes: we should then remove redundant text in the frame section. We may keep this text here, but I think merging it with Section 2.3 would avoid repeating this for all the PATH_* frames.

Author: I don't think the Frames section says this...? Which section do you mean by 2.3?

Author: @qdeconinck what do you mean?

Reviewer 3: I guess @qdeconinck is referring to [Section 7.4](https://datatracker.ietf.org/doc/html/draft-ietf-quic-multipath-13#name-path_new_connection_id-fram):

> Note that the NEW_CONNECTION_ID frame can only be used to issue or retire connection IDs for the initial path with Path ID 0.

Reviewer 2: Yeah, that was what @Yanmei-Liu said. What I meant is that we should say at one place that NEW_CONNECTION_ID, RETIRE_CONNECTION_ID and ACK frames are still valid and are equivalent to their PATH_* versions with Path ID 0. We can delegate this to a further issue, but we can just merge this for now.

### Issuing New Connection IDs

On non-initial paths
(i.e., Path ID different from 0), the initial path connection ID is provided by a
PATH_NEW_CONNECTION_ID frame with a sequence number value of 0.

> In order to let the peer open new paths, it is RECOMMENDED to proactively

Reviewer: No. Previous text is better. You may want to say that it is OK to send legacy NEW_CONNECTION_ID on path 0, but your text ends up saying that NEW_CONNECTION_ID frames cannot be used on path with number not equal to zero,
which is not what we agreed. How about:

The PATH_NEW_CONNECTION_ID frame is used to issue new connection IDs for any path. The NEW_CONNECTION_ID frame MAY be used to provide new connection IDs for path number 0. Connection ID sequence numbers start at 0.

Author: This is not a meant as a replacement. I simply moved this sentence up (see removed lines 547, 548, and 579 below to find the original sentence). And I removed the first sentence because I found it redundant with the text above (see lines 523-527) and also text in the frames section.

However, we can maybe also just remove this sentence as well as we say in line 528 already this:
"Each Path ID has its own connection ID sequence number space whose initial value is 0."

Reviewer: OK, let's remove the sentence.

Author: done

issue at least one Connection ID for each unused Path ID up to the
minimum of the peer's and the local Maximum Path ID limits.

> If for any reasons an endpoints does not want to issue connection IDs for all

Reviewer 2: ```suggestion
If for any reason an endpoint does not want to issue connection IDs for all
```

unused Path ID, it SHOULD NOT introduce discontinuity
in the issuing of Path IDs as path initiation
requires available connection IDs for the same Path ID on both sides. For instance,
if the maximum Path ID limit is 2 and the endpoint wants to provide connection IDs
for only one Path ID inside range \[1, 2\], it should select Path ID 1 (and not Path
ID 2).

Similarly, endpoints SHOULD consume Path IDs in a continuous way, i.e., when
creating paths. However, endpoints cannot expect to receive new connection IDs

> or path initiation attempts with in order use of Path IDs

Reviewer 2: ```suggestion
or path initiation attempts with in-order use of Path IDs
```

i.e., pick the last received token. To avoid issues when clients make the "wrong" choice,
a server SHOULD issue tokens that are capable of validating
any of the previously validated addresses. Further guidance on token usage can be
found in {{Section 8.1.3 of QUIC-TRANSPORT}}.

## Handling Connection IDs {#consume-retire-cid}

In order to identify a connection ID correctly when the multipath extension is used,
endpoints have to use the PATH_NEW_CONNECTION_ID and PATH_RETIRE_CONNECTION_ID frames
instead of the NEW_CONNECTION_ID and RETIRE_CONNECTION_ID frame to indicate
the respective Path ID together with associated sequence number
(see {{Section 5.1.1 of QUIC-TRANSPORT}}), at least for all paths with a Path ID other than 0.
Each Path ID has its own connection ID sequence number space whose initial value is 0.

Endpoints can also use PATH_NEW_CONNECTION_ID and
PATH_RETIRE_CONNECTION_ID for the initial path with Path ID 0,
however, the use of NEW_CONNECTION_ID and RETIRE_CONNECTION_ID
is still valid and endpoints need to process these frames accordingly
as corresponding to Path ID 0.

### Issuing New Connection IDs

> On non-initial paths
> (i.e., Path ID different from 0), the initial path connection ID is provided by a
> PATH_NEW_CONNECTION_ID frame with a sequence number value of 0.

Author: ```suggestion
```

and therefore are unique for the scope of the connection,
regardless of the associated Path ID.

Endpoints MUST NOT issue new connection IDs with Path IDs greater than
the Maximum Path Identifier field in MAX_PATH_ID frames (see {{max-paths-frame}})
or the value of initial_max_path_id transport parameter if no MAX_PATH_ID frame was received yet.
Receipt of a frame with a greater Path ID is a connection error as specified

> in Section {{frames}}.

Reviewer 2: ```suggestion
in {{frames}}.
```

When an endpoint finds it has not enough available unused path identifiers,
it SHOULD either send a MAX_PATH_ID frame to increase the active path limit
(when limited by the sender) or a PATHS_BLOCKED frame

> (see Section {{paths-and-cids-blocked-frame}}) to inform the peer that

Reviewer 2: ```suggestion
(see {{paths-and-cids-blocked-frame}}) to inform the peer that
```

its current limit prevented the creation of the new path.

### Rotating and Retiring Connection IDs

{{Section 5.1.2 of QUIC-TRANSPORT}} indicates that an endpoint
can change the connection ID it uses to another available one
at any time during the connection. For the extension specified in
this document, endpoints MUST only rotate to another connection IDs associated
to the same Path ID. Use of a connection ID associated with
another Path ID will be considered as an attempt to open a new path instead.

> An endpoints is supposed to retire connection ID that are not used anymore,

Reviewer 2: ```suggestion
An endpoint is supposed to retire connection ID that are not used anymore,
```

Reviewer: I would say, "An endpoint is expected to retire connection IDs that are not used anymore,"

and the server is supposed to provide
replacements, as specified in {{Section 5.1.2. of QUIC-TRANSPORT}}.

> As such when receiving a PATH_RETIRE_CONNECTION_ID frame, an endpoint

Reviewer 2: ```suggestion
As such, when receiving a PATH_RETIRE_CONNECTION_ID frame, an endpoint
```

Reviewer: Might be better to just remove "As such,"

SHOULD provide new connection IDs for that path, if still open, using PATH_NEW_CONNECTION_ID frames.

While it it expected that the peer provides at least one unused connection ID
for all active paths using the PATH_NEW_CONNECTION_ID after retirement
of an old connection ID, an endpoint MAY send

> a PATH_CIDS_BLOCKED (see Section {{paths-and-cids-blocked-frame}})

Reviewer 2: ```suggestion
a PATH_CIDS_BLOCKED (see {{paths-and-cids-blocked-frame}})
```

if it wants to change the connection ID but no
unused connection ID for a that path is available. Further, an
endpoint MAY also send a PATH_CIDS_BLOCKED frame if it wants to
open a new path and has no connection IDs available for an unused
path ID even though the Maximum Path Identifier value would allow
for more paths.

Retirement of connection IDs will not retire the Path ID
that corresponds to the connection ID or any other path resources

> as the packet number space is associated with to the Path ID.

Reviewer 2: ```suggestion
as the packet number space is associated to the Path ID.
```

PATH_AVAILABLE indicates that a path is "available", i.e., it suggests to
the peer to use its own logic to split traffic among available paths.

PATH_BACKUP suggests that a path should only be used as backup i.e. that no traffic
should be sent on that path if another path is avaialble and usable.
If all established paths are indicated as backup paths, no guidance is provided about
which path should be used.

Similarly, if no frame indicating a path usage preference was received for a certain path,
the preference of the peer is unknown and the sender needs to decide based on it
own local logic if the path should be used.

If an endpoint starts using a backup path
because it has detected issues on the paths marked as "available", it is RECOMMENDED
to update its own path state signaling such that the peer avoids using the broken path.
An endpoint that detects a path breakage can also explicitly close the path

> by sending a PATH_ABANDON frame (see Setion {{path-close}}) in order to avoid

Reviewer 2: ```suggestion
by sending a PATH_ABANDON frame (see {{path-close}}) in order to avoid
```

that its peer keeps using it and enable faster switch over to a backup path.
If the endpoints do not want to close the path immediately, as connectivity
could be re-established, PING frames can potentially be used to quickly detect
connectivity changes and switch back in a timely way.

The PATH_AVAILABLE and PATH_BACKUP frames share a common, per-path sequence number space
to detect and ignore outdated information, as further described in {{path-backup-available-frame}}.
This is needed as they might arrives in different orders,
e.g if sent via different paths.

## Path Close {#path-close}

At any time in the connection, each endpoint can decide to

> abandon a paths, for example following changes in local

Reviewer: "abandon a path," (singular)

Author: ```suggestion
abandon a path, for example following changes in local
```

Each endpoint manages the set of paths that are available for
transmission. At any time in the connection, each endpoint can decide to
abandon one of these paths, for example following changes in local
connectivity or local preferences. After an endpoint abandons
a path, the peer can expect to not receive any more packets on
that path.

Note that other explicit closing mechanisms of {{QUIC-TRANSPORT}} still
apply on the whole connection. In particular, the reception of either a
CONNECTION_CLOSE ({{Section 10.2 of QUIC-TRANSPORT}}) or a Stateless
Reset ({{Section 10.3 of QUIC-TRANSPORT}}) closes the connection.
At any time in the connection, each endpoint can decide to
abandon a paths, for example following changes in local
connectivity or local preferences.
An endpoint that wants to abandon a path MUST explicitly
close the path by sending a PATH_ABANDON frame (see {{path-abandon-frame}}).
This is true whether the decision to close the path results
from implicit signals such as an idle time-out (see {{idle-time-close}})
or packet losses as well as for any other reason such as management
of local resources.

An endpoint that wants to close a path MUST explicitly

> terminate the path by sending a PATH_ABANDON frame (see {{path-abandon-frame}}).

Reviewer: Can we keep that text? It says something very clearly -- that there is no other way to close a path. That's the main message of the "close path" section.

> does not indicate path abandonment (see further Section {{consume-retire-cid}}).

Reviewer 2: ```suggestion
does not indicate path abandonment (see further {{consume-retire-cid}}).
```

> not only the path that is intended to be closed.

Reviewer 2: ```suggestion
not only on the path that is intended to be closed.
```

> {{spurious-stateless-reset}}, and helps to acknowledge any

Reviewer 2: ```suggestion
{{spurious-stateless-reset}}, and helps acknowledge any
```

> potentially reordered, outstanding packets from the peer (see Section {ack-after-abandon}).

Reviewer 2: ```suggestion
potentially reordered, outstanding packets from the peer (see {{ack-after-abandon}}).
```

> PATH_ABANDON frame (see Section {{path-initiation}}). An endpoint may also decide

Reviewer 2: ```suggestion
PATH_ABANDON frame (see {{path-initiation}}). An endpoint may also decide
```

it MAY close the connection, but SHOULD wait at least 3 PTOs after
sending the PATH_ABANDON frame.

After a path is abandoned, the Path ID MUST NOT be reused
for new paths, as the Path ID is part of the nonce calculation {{multipath-aead}}.

If a PATH_ABANDON frame is received for the only open path of a QUIC
connection, the receiving peer SHOULD send a CONNECTION_CLOSE frame
and enter the closing state. Alternatively, a client MAY instead try to open a new path, if
available, and only initiate connection closure if path validation fails

> or a CONNECTION_CLOSE frame is received from the server. Similarly

Reviewer 2: ```suggestion
or a CONNECTION_CLOSE frame is received from the server. Similarly,
```

### Avoiding Spurious Stateless Resets {#spurious-stateless-reset}

Due to network delays, packets sent on an abandoned path can
arrive well after the connection IDs have been retired.
If not recognized as bound to the local

> connection, such packet trigger the peer to send a Stateless Reset

Reviewer 2: ```suggestion
connection, such packet triggers the peer to send a Stateless Reset
```

packet. The rule to "retain knowledge of connection ID for 3 PTO
after receiving a PATH_ABANDON"
is intended to reduce the risk of sending such spurious stateless
packets, but it cannot completely avoid that risk.

{{Section 10.3 of QUIC-TRANSPORT}} specified that the Stateless Reset Tokens
associated with retired connection IDs cannot be used to identify Stateless Reset packets.
The immediate retirement of connection IDs received from the peer for an abandoned
path guarantees that spurious Stateless Reset packets
sent by the peer will not cause the closure of the QUIC connection.

### Handling PATH_ACK for Abandoned Paths {#ack-after-abandon}

When an endpoint sends a PATH_ABANDON frame, there may
still be some packets in transit from the peer.
Further, if an endpoint receives a PATH_ABANDON frame, it may still receive

> reordered packets on the abanded path. Endpoints SHOULD

Reviewer 2: ```suggestion
reordered packets on the abandoned path. Endpoints SHOULD
```

> promptly send PATH_ACK frames for all unacknowledge packets received on

Reviewer 2: ```suggestion
promptly send PATH_ACK frames for all unacknowledged packets received on
```

> an abondoned path if path state is still retained to do so.

Reviewer 2: ```suggestion
an abandoned path if path state is still retained to do so.
```

PATH_ACK frames have to be sent on a different path than the path being abandoned
after sending the PATH_ABANDON frame as connection IDs are immediately retired.

When an endpoint finally deletes all state associated with the path,
the packets sent over the path and not yet acknowledged MUST be considered lost.
PATH_ACK frames received with an abandoned Path ID are silently ignored,

> as specified in Section {{frames}}.

Reviewer 2: ```suggestion
as specified in {{frames}}.
```
